garage/plotter/plotter.py

```diff
@@ -1,88 +1,153 @@
-                policy.set_param_values(param_values)
-                rollout(
-                    env,
-                    policy,
-                    max_path_length=max_length,
-                    animated=True,
-                    speedup=5)
-            else:
-                if max_length:
+                elif Op.UPDATE in msgs:
+                    env, policy = msgs[Op.UPDATE].args
+                elif Op.DEMO in msgs:
+                    param_values, max_length = msgs[Op.DEMO].args
+                    policy.set_param_values(param_values)
+                    initial_rollout = False
                     rollout(
                         env,
                         policy,
                         max_path_length=max_length,
                         animated=True,
                         speedup=5)
-    except KeyboardInterrupt:
-        pass
-
-
-def _shutdown_worker():
-    if process:
-        queue.put(['stop'])
-        queue.close()
-        process.join()
-
-
-def init_worker():
-    global process, queue
-    queue = Queue()
-    process = Thread(target=_worker_start) if (
-        'Darwin' in platform.platform()) else Process(target=_worker_start)
-    process.daemon = True
-    process.start()
-    atexit.register(_shutdown_worker)
-
-
-def init_plot(env, policy):
-    global process, queue
-    if not (process and queue):
-        init_worker()
-
-    # Needed in order to draw glfw window on the main thread
-    if ('Darwin' in platform.platform()):
-        rollout(env, policy, max_path_length=np.inf, animated=True, speedup=5)
-
-    queue.put(['update', env, policy])
+                else:
+                    if max_length:
+                        rollout(
+                            env,
+                            policy,
+                            max_path_length=max_length,
+                            animated=True,
+                            speedup=5)
+        except KeyboardInterrupt:
+            pass
+
+    def shutdown(self):
+        if self._process.is_alive():
+            self._queue.put(Message(op=Op.STOP, args=None, kwargs=None))
+            self._queue.close()
+            self._process.join()
+
+    @staticmethod
+    def set_enable(enable):
+        """Set the plot enabling according to the run_experiment function.
+        """
+        Plotter.enable = enable
+
+    @staticmethod
+    def status(plot_algo_status):
+        """Use this method as a guard for plot enabling.
+
+        This guard method considers the plot enabling defined in the
+        function run_experiment.
+
+        Parameters
+        ----------
+        plot_algo_status : boolean
+            The plot enabling defined in the algorithm class.
+
+        Returns
+        -------
+            The plot enabling of the algorithm if the run_experiment was not
+            called, and the plot enabling of the run_experiment otherwise.
+
+        """
+        if Plotter.enable is None:
```

Comment:
in general you don't want to distinguish between NoneType and False in application logic. You are overloading None to mean "nobody has set this yet." 

I think you can get equivalent behavior by initializing Plotter.enable to true. Then if either the algorithm or run_experiment calls `Plotter.disable()` the plotter is permanently disabled (there would not need to be a Plotter.enable() API, but there could be).